ubccr/xdmod
# Refactor some of the integration tests.

## Description



This refactors some of the integration tests to make the code simpler, less duplicated, and more consistent. Specifically,
- Change `TokenHelper` to be a test class called `TokenAuthTest` that can be extended to run a suite of token authentication tests on an endpoint: empty token, malformed token, invalid token, valid token (one or more tests), expired token, and revoked token. This improves the test coverage of the `Tokens::authenticateToken()` method.
- Move the `makeRequest()` method from `TokenHelper` into a new `requestAndValidateJson()` method in `BaseTest`. This provides a method that any test can reuse.
- When expiring tokens, set a time in the past rather than the current time. This fixes a bug in which a test could run fast enough that the difference between the current and past times was too insignificant to be detected.
- Move API token endpoint tests from `UserControllerProviderTest` into the test classes for the controllers that implement those endpoints, `MetricExplorerTest` and `WarehouseExportControllerProviderTest`. This keeps tests closer to the endpoints they are testing.
- Use the existing `Configuration` functionality to parse JSON test artifact files, adding an `IntegrationTestConfiguration` class that allows files to refer to paths relative to the integration test artifacts root directory and allow JSON schemas to have references to keys in the same document (e.g., `"$ref": "#/foo"`).
- Add a `dataProvider` to `BaseTest` that simply provides the list of roles, which is sufficient for many tests.
- Move JSON file loading and validation into a `validateJsonAgainstFile()` method since the tests that called the `validateJson` method all involved loading a file first. This reduces duplicated code.
- Add the [`swaggest/php-json-schema`](https://github.com/swaggest/php-json-schema) library to enable validating JSON using JSON Schema Draft 7. The existing [`justinrainbow/json-schema`](https://github.com/justinrainbow/json-schema) library only supports up through JSON Schema Draft 4. For example, the `contains` array property can now be validated; it was introduced in JSON Schema Draft 6.
- Make the expected "unauthorized" error code be only 401 since el7 does not actually produce 400.
- Move `getTestFiles()` into `BaseTest` to avoid multiple files reimplementing it.
- Make the `getRealms()` method of `BaseTest` be `protected` instead of `public` since it doesn't need to be called outside of its class hierarchy.
- Rename classes to match the names of the classes they are testing (e.g., `WarehouseExportControllerTest` becomes `WarehouseExportControllerProviderTest`).
- Fix warnings reported by `phpcs`.

## Motivation and Context


In addition to making the integration test code simpler, less duplicated, and more consistent, these changes also provide a foundation for additional integration tests to be added in https://github.com/ubccr/xdmod/pull/1718, https://github.com/ubccr/xdmod/pull/1720, and https://github.com/ubccr/xdmod/pull/1721.

## Tests performed



To verify the correct tests ran producing the correct results, I did the following:
- For both of the Docker images `tools-ext-01.ccr.xdmod.org/xdmod-10.0.0:centos7.9-0.6` and `tools-ext-01.ccr.xdmod.org/xdmod10.0:rockylinux8.5-0.5`:
    - Spin up a Docker container from the image.
    - Upgrade to the latest Open XDMoD 10.5 version.
    - Run the `~/bin/services start` command.
    - In a cloned copy of https://github.com/ubccr/xdmod, switch to the `xdmod10.5` branch.
    - Prepare for the integration tests (as done in `.circleci/config.yml`):
        - `mv ./configuration/organization.json{,.old}`
        - `mv ./configuration/portal_settings.ini{,.old}`
        - `cp {/opt/xdmod-10.5.0/etc,./configuration}/portal_settings.ini`
        - `cp {/opt/xdmod-10.5.0/etc,./configuration}/organization.json`
    - Change to the `tests/integration` directory and run:
        - `XDMOD_REALMS=jobs,storage,cloud ../../vendor/bin/phpunit --testsuite default --group UserAdminTest.createUsers --debug > before.txt`
        - `XDMOD_REALMS=jobs,storage,cloud ../../vendor/bin/phpunit --testsuite default --exclude-group UserAdminTest.createUsers --debug >> before.txt`
    - Merge in the changes from this PR.
    - Run:
        - `XDMOD_REALMS=jobs,storage,cloud ../../vendor/bin/phpunit --testsuite default --group UserAdminTest.createUsers --debug > after.txt`
        - `XDMOD_REALMS=jobs,storage,cloud ../../vendor/bin/phpunit --testsuite default --exclude-group UserAdminTest.createUsers --debug >> after.txt`
    - Compare the `before.txt` and `after.txt` files and make sure the differences are correct.

## Checklist:
